### DISCUSSION TITLE
Is it possible/neccessary to specify the position of the layer swapped within the smart layer itself?

### DISCUSSION THREAD

**Discussion by DigitalBeer** [joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script#26] [Q&A]
Hello Joonas,

Firstly, thank you, for making this available and engaging with the questions people have asked, I've been having great fun tinkering with this script engine on my own personal projects, it's super fun.

I've had a look through documentation and all the Q&A on GitHub and I can't find a definitive answer, I'm hoping you can tell me if I'm getting something wrong.

I have a number of psd files for art frame mock-ups, inside is a smart object and within that smart object are a lot of effect layers, shadows, textures etc. When I'm running the script the exported jpg of my image doesn't have any of those effect layers applied. I am assuming that this is because "The old switcheroo" 👍 is happening as the top layer within the smart object? I tried the "showLayers" function and specified each of the effect layers in the array but that didn't seem to have an effect.

So my question is ultimately, is it possible to specify the position of the layer swapped within the smart layer itself, or should that not actually be an issue and my issue lies somewhere else?

Cheers!

**Comment by joonaspaakko** [comment]
I'm not sure if I understood the issue, but I'm thinking maybe `nestedTarget`. You give the script a smart object `target` like always, then you can make an additional smart object inside it and target that layer using the `nestedTarget`, which will then be where the input files are placed:

> In the picture the floating window on the left is the `target` smart object contents and in the layers panel you can see the `nestedTarget`.

![image](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/assets/1164476/6395eb88-4948-4151-b6aa-286868818764)

In all of my examples of the nested target, it's smaller than the regular target, but in your case it should be the same size as its parent target since if I understood correctly, you only need it in order to place the input files inside the target smart object at a very specific layer position.

I have two examples in the repository that use it: https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/8874ffaad31dd9199a422caed3a4f5f2c84c528a/Examples/example-1/Batch%20replace%20-%20example%201.jsx#L57-L67

This is perhaps easier to understand with the target names. The target smart object holds a browser mockup and the nested target holds the browser viewport contents...
https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/8874ffaad31dd9199a422caed3a4f5f2c84c528a/Examples/example-2/Batch%20replace%20-%20example%202.jsx#L37-L44

**Comment by DigitalBeer** [comment]
Hello Joonas, thank you for your time and suggestion.

I will try to explain the original issue better so as not to waste your time, when I was referring to layer position I meant it's position relative to other layers e.g. top layer.

Here is the original mockup
![Screenshot 2023-08-13 162600](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/assets/130303270/b9fda70a-ebab-4e52-9334-36ab85665d42)


Here is the contents of the smart object, as you can see, all the effect layers that are specific to the poster texture are in here.
![Screenshot 2023-08-13 162614](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/assets/130303270/c7d378b2-cce2-45a5-94ab-a4af5e99eaf8)


When I run the script using any input image, it comes out looking like this, the image is fine but you can see that it is is missing all the paper texture like the folds and tears and shadows from within the smart object.
![image-from-rawpixel-id-421171-jpeg - UK ISO 210 International Frame Mockup - poster wall urban poster mockup - 1](https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/assets/130303270/4ed13cd2-f359-4161-b250-6ae8fddd5c44)


Does that make more sense now? Is nestedTarget still the correct approach?

Thank you

**Reply by joonaspaakko** [reply]
Yea, it's the same exact scenario as the billboard mockup in example 2

https://github.com/joonaspaakko/Batch-Mockup-Smart-Object-Replacement-photoshop-script/blob/8874ffaad31dd9199a422caed3a4f5f2c84c528a/Examples/example-2/Batch%20replace%20-%20example%202.jsx#L37-L44

You can convert the `Placeholder` layer into a smart object and that's your `nestedTarget`:

```js
target: '@UK ISO 210 International',
nestedTarget: 'Placeholder',
```

**Comment by DigitalBeer** [comment]
Hi Joonas, you were absolutely right, thank you for clarifying that for me, I have done exactly what you explained and it's working perfectly now. Thank you very much for your time and knowledge. You deserve a🥇

Cheers 👍
